include/pybind11/detail/function_view.h

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2021 The Pybind Development Team.
+// All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+#pragma once
+
+#include "common.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <typename Signature>
+class function_view;
+
+// Type-erased unowned reference to a function/function object.
+//
+// This is similar to `std::function`, but is much lighter weight (size of two
```

Comment:
With 64-bit ubuntu & clang++:
* `sizeof(py::detail::type_record)` is 15 pointers.
* `sizeof(std::function<void()>)` is 4 pointers.

That makes me think the optimization is 17 pointers vs 19 pointers.

I wonder about the size of the generated machine code: if `std::function` is reused, could that be a net saving, even though sizeof(std::function) is 4 vs sizeof(function_view) is 2?

I noticed this existing comment in attr.h:
```
/// Special data structure which (temporarily) holds metadata about a bound class
struct type_record {                                                            
```

IIUC, if we reuse `std::function` we'd have the same functionality with less code ("less is more").

Do we have measurements that could justify adding this extra code?

Reply:
The space usage isn't really an issue since `type_record` is temporary and only lives on the stack.

We could definitely use `std::function` --- that would be just fine.  I avoided it in part because pybind11 currently doesn't use `std::function` except in the separate, not default-included header, `pybind11/functional.h`.

I don't think there is really any situation in which `std::function` is more efficient or faster to compile or produces smaller generated code than `function_view`:

https://godbolt.org/z/57rMnnjTT

But in any case, I am happy to change to `std::function` if that is preferred.  Certainly if we don't anticipate any further uses of `function_view` (basically equivalent to `absl::FunctionRef` by the way) within pybind11 then it is probably not worth adding it for this purpose.

Reply:
> We could definitely use std::function --- that would be just fine.

Could you please make that change? I think then this PR is good to go in.

Reply:
Done.

Reply:
I'm not sure if there are currently other places where it would be useful at the moment.  I don't see any uses of `std::function` outside of pybind11/functional.h.

Just let me know if it should be brought back or not.